src/backend/src/heap_state/btc_user_pending_tx_state.rs

```diff
@@ -0,0 +1,423 @@
+            max_pending_transactions: max_pending_txs.unwrap_or(MAX_PENDING_TRANSACTIONS),
+        }
+    }
+
+    /// Returns the pending transactions of a specific principal per address.
+    #[allow(dead_code)]
+    pub fn get_pending_transactions(
+        &self,
+        principal: &Principal,
+        address: &str,
+    ) -> &Vec<StoredPendingTransaction> {
+        static EMPTY_VEC: Vec<StoredPendingTransaction> = Vec::new();
+        self.pending_transactions_map
+            .get(principal)
+            .map_or(&EMPTY_VEC, |map| map.get(address).unwrap_or(&EMPTY_VEC))
+    }
+
+    /// Adds a pending transaction for a specific principal and address.
+    /// It has a limit of storable transactions set on init.
+    #[allow(dead_code)]
+    pub fn add_pending_transaction(
+        &mut self,
+        principal: Principal,
+        address: String,
+        new_transaction: StoredPendingTransaction,
+    ) -> Result<(), String> {
+        if let Some(address_map) = self.pending_transactions_map.get_mut(&principal) {
+            if let Some(list) = address_map.get_mut(&address) {
+                if list.len() >= self.max_pending_transactions {
+                    return Err("Maximum pending transactions reached".to_string());
+                }
+                list.push(new_transaction);
+            } else {
+                let list: Vec<StoredPendingTransaction> = vec![new_transaction];
+                address_map.insert(address, list);
+            }
+        } else {
+            let mut address_map: PendingTransactionsMap = HashMap::new();
+            let list: Vec<StoredPendingTransaction> = vec![new_transaction];
+            address_map.insert(address, list);
+            self.pending_transactions_map.insert(principal, address_map);
+        }
+        Ok(())
+    }
+
+    /// Prunes pending transactions for a specific principal.
+    /// A pending transaction can be pruned for two reasons:
+    /// - Transaction is older than 1 day.
+    ///   We consider that if a pending transaction is older than one day
+    ///   it means it failed and we can free to utxos to be used again.
+    /// - None of the transaction's utxos are present in the current utxos list.
+    ///   We use the pending transactions to avoid double spending.
+    ///   Once we know that a utxos is not available, we can remove the pending transaction.
+    ///   Normally, all utxos of a pending transaction should be present or not.
+    ///   Partial presence could happen if the utxos of a pending transaction were not really used in the transaction.
+    ///   We don't remove in partial presence because, in the end, partial presence will be temporary for one day.
+    #[allow(dead_code)]
+    pub fn prune_pending_transactions(
+        &mut self,
+        principal: Principal,
+        current_utxos: &[Utxo],
+        now_ns: u64,
+    ) {
+        if let Some(address_map) = self.pending_transactions_map.get_mut(&principal) {
```

Comment:
Do I see correctly that the number of addresses per principal is unbounded?

Reply:
Fixed!